Assets/Mirror/Components/InterestManagement/Match/MatchInterestManagement.cs

```diff
@@ -7,129 +7,118 @@ namespace Mirror
-            // Add the current match to dirtyMatches for Update to rebuild it.
-            dirtyMatches.Add(networkMatchId);
+            matchCount = (byte)matchObjects.Count;
         }
 
         [ServerCallback]
-        public override void OnDestroyed(NetworkIdentity identity)
+        void RebuildMatchObservers(Guid matchId)
         {
-            // Don't RebuildSceneObservers here - that will happen in Update.
-            // Multiple objects could be destroyed in same frame and we don't
-            // want to rebuild for each one...let Update do it once.
-            // We must add the current match to dirtyMatches for Update to rebuild it.
-            if (lastObjectMatch.TryGetValue(identity, out Guid currentMatch))
-            {
-                lastObjectMatch.Remove(identity);
-                if (currentMatch != Guid.Empty && matchObjects.TryGetValue(currentMatch, out HashSet<NetworkIdentity> objects) && objects.Remove(identity))
-                    dirtyMatches.Add(currentMatch);
-            }
+            foreach (NetworkMatch networkMatch in matchObjects[matchId])
+                if (networkMatch.netIdentity != null)
+                    NetworkServer.RebuildObservers(networkMatch.netIdentity, false);
         }
 
-        // internal so we can update from tests
+        // called by NetworkMatch.matchId setter
         [ServerCallback]
-        internal void Update()
+        internal void OnMatchChanged(NetworkMatch networkMatch, Guid oldMatch)
         {
-            // for each spawned:
-            //   if match changed:
-            //     add previous to dirty
-            //     add new to dirty
-            foreach (NetworkIdentity identity in NetworkServer.spawned.Values)
-            {
-                // Ignore objects that don't have a NetworkMatch component
-                if (!identity.TryGetComponent(out NetworkMatch networkMatch))
-                    continue;
+            // Mark new/old matches as dirty so they get rebuilt
 
-                Guid newMatch = networkMatch.matchId;
-                if (!lastObjectMatch.TryGetValue(identity, out Guid currentMatch))
-                    continue;
+            // Guid.Empty is never a valid matchId
+            if (networkMatch.matchId != Guid.Empty)
+                dirtyMatches.Add(networkMatch.matchId);
 
-                // Guid.Empty is never a valid matchId
-                // Nothing to do if matchId hasn't changed
-                if (newMatch == Guid.Empty || newMatch == currentMatch)
-                    continue;
+            dirtyMatches.Add(networkMatch.matchId);
 
-                // Mark new/old matches as dirty so they get rebuilt
-                UpdateDirtyMatches(newMatch, currentMatch);
+            // This object is in a new match so observers in the prior match
+            // and the new match need to rebuild their respective observers lists.
```

Comment:
Comment without any code?

Reply:
describes what needs to be done in the steps to follow, since two methods were merged into that method at that point. Keeping the comments for now.